fsharp-backend/src/LibExecution/Interpreter.fs

```diff
@@ -48,6 +49,279 @@ let rec eval' (state : ExecutionState) (st : Symtable) (e : Expr) : DvalTask =
+    let rec patternPasses dv pattern : bool =
+      match pattern with
+      | PInteger (_, i) -> dv = DInt i
+      | PBool (_, b) -> dv = DBool b
+      | PCharacter (_, c) -> dv = DChar c
+      | PString (_, str) -> dv = DStr str
+      | PFloat (_, v) -> dv = DFloat v
+      | PNull (_) -> dv = DNull
+      | PVariable (_, _) -> not (Dval.isFake dv)
+      | PBlank (_) -> false
+      | PConstructor (_, name, args) ->
+        match (name, args, dv) with
+        | "Nothing", [], DOption None -> true
+
+        | "Just", [ p ], DOption (Some _v)
+        | "Ok", [ p ], DResult (Ok _v)
+        | "Error", [ p ], DResult (Error _v) -> patternPasses dv p
+
+        | "Nothing", [], _ -> false
+        | _ -> false
+
+    let tracePattern onExecutionPath dv pattern : List<string * Dval> =
+      let traceFn (id, dv) = state.tracing.traceDval onExecutionPath id dv
+      let traceIncomplete id = state.tracing.traceDval false id (incomplete id)
+
+      let rec r dv pattern =
+        match pattern with
+        | PInteger (id, i) ->
+          traceFn (id, DInt i)
+          []
+        | PBool (id, b) ->
+          traceFn (id, DBool b)
+          []
+        | PCharacter (id, c) ->
+          traceFn (id, DChar c)
+          []
+        | PString (id, s) ->
+          traceFn (id, DStr s)
+          []
+        | PFloat (id, f) ->
+          traceFn (id, DFloat f)
+          []
+        | PNull (id) ->
+          traceFn (id, DNull)
+          []
+        | PBlank (id) ->
+          traceFn (id, incomplete id)
+          []
+        | PVariable (id, varName) ->
+          traceFn (id, dv)
+          [ (varName, dv) ]
+        | PConstructor (id, name, args) ->
+          match (name, args, dv) with
+          | "Nothing", [], _ ->
+            traceFn (id, DOption None)
+            []
+
+          | "Just", [ p ], DOption (Some v)
+          | "Ok", [ p ], DResult (Ok v)
+          | "Error", [ p ], DResult (Error v) when patternPasses v p ->
+            traceFn (id, dv)
+
+            r v p
+
+          | _name, subPatterns, _ ->
+            traceIncomplete id
+
+            subPatterns
+            |> List.map Pattern.toID
+            |> List.iter (fun pId -> traceFn (pId, incomplete pId))
+
+            []
+
+      r dv pattern
+
+    uply {
+      // The value we're matching against
+      let! matchVal = eval state st matchExpr
+
+      let stWithNewVars newVars = Map.mergeFavoringRight st (Map.ofList newVars)
+
+      let mutable foundMatch : Option<Dval> = None
+
+      do!
+        cases
+        |> Ply.List.iterSequentially (fun (pattern, rhsExpr) ->
+          let passes = patternPasses matchVal pattern
+
+          match foundMatch, passes, state.tracing.realOrPreview with
+          // Found a match - persist traces and set the result
+          | None, true, _ ->
+            uply {
+              let newDefs = tracePattern state.onExecutionPath matchVal pattern
```

Comment:
The split between trace pattern pattern passes feels a little bit off to me, in a few ways:
- tracePattern fetches the matched values, which feels outside its scope (at least, given its name)
- the logic of whether something should match is repeated twice

I think if pattern passes returned the list of ids and Dvals, then tracing could just be iterating through that list, with no need to repeat the logic. I'm not sure if that would work, just an idea.

Reply:
Take another look now :)

Reply:
Yeah, this is much nicer. Is it possible to only check the pattern if we haven't already matched?

Reply:
Done now. WDYT? I'm not a huge fan of the `caseIdx` use, but it achieves all of our goals, I think, and it's a small enough surface.

Reply:
I thought the old version was fine in general, and just required moving things around a bit. There's definitely a lot more moving parts in the new version, so definitely like it less.

Reply:
When I went towards "only check the pattern if we haven't already matched," every solution I tried seemed bad in some way - I'll give it another short round - maybe I missed something.

Reply:
Yeah I think so

Reply:
Check again - looks better to me.

I'm not sure why we needed the Ply.List.iterSequentially - I think this is fine with a simple `for ... in` loop?